File: JavaScript/regularExpressionMatching.js
function isMatch(s, p) {
  const m = s.length;
  const n = p.length;

  // Create a 2D DP table
  const dp = new Array(m + 1);
  for (let i = 0; i <= m; i++) {
    dp[i] = new Array(n + 1).fill(false);
  }

  // Base case: an empty pattern matches an empty string
  dp[0][0] = true;

  // Handle the case when s is empty but p has '*'
  for (let j = 1; j <= n; j++) {
    if (p[j - 1] === '*') {
      dp[0][j] = dp[0][j - 2];
    }
  }

  // Fill the DP table
  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      if (p[j - 1] === s[i - 1] || p[j - 1] === '.') {
        dp[i][j] = dp[i - 1][j - 1];
      } else if (p[j - 1] === '*') {
        dp[i][j] = dp[i][j - 2];
        if (p[j - 2] === s[i - 1] || p[j - 2] === '.') {
          dp[i][j] |= dp[i - 1][j];
        }
      }
    }
  }

  return dp[m][n];
}

File: C#/regularExpressionMatching.cs
public class Solution {
    public bool IsMatch(string s, string p) {
        int m = s.Length;
        int n = p.Length;
        
        // Create a 2D DP table
        bool[,] dp = new bool[m + 1, n + 1];
        
        // Base case: an empty pattern matches an empty string
        dp[0, 0] = true;
        
        // Handle the case when s is empty but p has '*'
        for (int j = 1; j <= n; j++) {
            if (p[j - 1] == '*') {
                dp[0, j] = dp[0, j - 2];
            }
        }
        
        // Fill the DP table
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (p[j - 1] == s[i - 1] || p[j - 1] == '.') {
                    dp[i, j] = dp[i - 1, j - 1];
                } else if (p[j - 1] == '*') {
                    dp[i, j] = dp[i, j - 2];
                    if (p[j - 2] == s[i - 1] || p[j - 2] == '.') {
                        dp[i, j] |= dp[i - 1, j];
                    }
                }
            }
        }
        
        return dp[m, n];
    }
}


We have two loops, one nested inside the other. The outer loop is controlled by the variable i, 
and it goes from 1 to m (the length of string s). The inner loop is controlled by the variable j, 
and it goes from 1 to n (the length of pattern p).

Inside these loops, we perform some checks and calculations. Here's what happens:

We check if the character at position j - 1 in the pattern p is equal to the character at 
position i - 1 in the string s, or if it is a dot ('.'). If it is, it means we have a match, and we set 
the corresponding value in the dp table to be the same as the value in the dp table for the previous 
positions i - 1 and j - 1. This is represented by dp[i][j] = dp[i - 1][j - 1].

If the character at position j - 1 in the pattern p is an asterisk ('*'), we have a special case. We set
the current dp value to be the value two positions before in the same row, which is dp[i][j - 2]. This accounts 
for the scenario where the asterisk represents zero occurrences of the previous character in the pattern.

Additionally, if the character before the asterisk at position j - 2 in the pattern p is equal to the character 
at position i - 1 in the string s, or if it is a dot ('.'), we update the current dp value by performing a logical 
OR operation with the value at the previous row and the same column, dp[i - 1][j]. This accounts for the scenario where 
the asterisk represents one or more occurrences of the previous character in the pattern.

Finally, the function returns the value in the bottom-right cell of the dp table, which represents whether the entire 
pattern matches the entire string.

I hope this explanation helps clarify the code for a 5th grader! Let me know if you have any more questions.




